Build otherlibs (and other installed artifacts) using our stdlib

dkalinichenko-js opened the pull request:
Split the last stage of our compiler build into two: `compiler` and `main`. `compiler` mimics `boot-compiler`, but uses our standard library instead of the upstream version to build the targets. `main` builds the rest of the installation using that compiler, reusing artifacts already built.

This is necessary to compile `Stdlib_alpha.Or_null` in bytecode correctly: the upstream standard library can't marshal null constants. This is also helpful for supporting `float32` in bytecode.

Note that `unix` is still compiled using `boot-compiler`, since `compiler` depends on `unix`.

It's complicated to add a test for this, since `or_null` in bytecode is not implemented yet, but I successfully built my WIP bytecode branch with this change. I also inspected `_build/log` to ensure it uses the right compiler.

cc @TheNumbat

mshinwell commented:
I'll look at this - there have been some problems with 5.2 which might be related.

dkalinichenko-js commented:
I'm looking into the CI failures, can't reproduce them on my machine so far.

mshinwell commented:
@dkalinichenko-js can you please get CI to pass on this PR?

dkalinichenko-js commented:
Seems like Dune clears the `_build/main` default directory between builds in the CI for some reason, while it doesn't locally. I fixed this by hardlinking the resulting compiler instead of symbolic linking. This is not ideal, since we want Dune to reuse already built artifacts in `_build/main`, but should be sufficient for testing.

mshinwell commented:
Can we try to do something to avoid hardlinking tricks, which can be fragile and difficult to understand?

mshinwell commented:
(I think the change in behaviour depends on the Dune version, we've encountered this elsewhere iirc)

dkalinichenko-js commented:
I'll try that, agree it's fragile and bad.

dkalinichenko-js commented:
I updated the PR to build actual final targets instead of boot targets, because boot targets don't depend on our version of unix and it's very tricky to make them to. I also removed hardlinking in favor of copying. I'm not quite happy with this, since we perform duplicate work if Dune clears `_build/main` between builds, but if the impact on build times is small this should unblock various bytecode changes.

mshinwell commented on Makefile.common-jst:
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
-	@touch _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/dynlink.cmxa
+# Create [dynlink.cmxa] for 4.14 and [dynlink/dynlink.cmxa] for 5.0
+	mkdir -p _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/dynlink
+	@touch _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/{,dynlink/}dynlink.cmxa
```
Should this change be in a separate PR?

dkalinichenko-js replied:
Previously, Dune assumed the 4.14 directory structure for the `boot-compiler`. Now, it also assumes the 5.0 directory structure for the `compiler` we use to build otherlibs, so we need this change to build everything.

mshinwell commented on Makefile.common-jst:
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
 
 compiler: runtime-stdlib
+# Build the same target as in [boot-compiler], but with our runtime stdlib.
+# Uses the [main] context to avoid duplicating work.
```
I think this "Uses the..." comment is in the wrong place.

mshinwell commented on Makefile.common-jst:
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
+# incorrectly, using the upstream stdlib. This miscompiles code using new constants
+# like null pointers or float32 values.
+# However, otherwise [boot-compiler] produces identical artifacts to [compiler],
+# and it's safe-ish to reuse the [main] context again.
```
We need a better argument than "safe-ish".  Which artifacts are required from the context in which the final compiler is built in order to build the otherlibs (minus `unix`)?  I'm wondering if we can use a separate context for this last part.

dkalinichenko-js replied:
Moved the entire `dune_compiler_context` to its own directory. This means we don't reuse any artifacts, potentially slowing down build times but ensuring we do not miscompile anything. I believe, since `main` was already cleared between builds in the CI, this change should not introduce any extra delay there.
